src/main/java/com/blog/stream/pagination/PageSpliterator.java

```diff
@@ -1,108 +1,189 @@
     }
 
     /**
-     * Attempt to split the spliterator. This is easy as by definition the result set is defined as paged buckets
-     *
-     * If the stream is run in parallel the spliterator will be split as below
-     * (4 queries for a given pageSize/count, top row is page number)
-     * |   0    |   1    |   2    |   3    |
-     * | Parent |        |        |        |  No Splits
-     * | Child  | Parent |        |        |  First Split
-     * | Child  | Child  | Parent |        |  Second Split
-     * | Child  | Child  | Child  | Parent |  Third Split
-     *                                        No more splits (return null)
-     *
-     *  If run sequentially, try split will not be called and the tryAdvance handles paging
-     * @return Child spliterator whose role is to iterate over one set and die.
+     * Construct a PageSpliterator for parallel with prefetch to get more accurate page count (must have totalPagesExtractor)
+     * @param pageSize size of each page
+     * @param pageFetcher function to fetch a page of type P given both page and page size
+     * @param itemExtractor function to extract list of item type T from page object of type P
+     * @param totalPagesExtractor function to extract page count from page of type P
+     * @param <P> page object type
+     * @param <T> item type
+     * @return a PreFetchPageSpliterator configured for prefetch
      */
+    public static <P, T> PageSpliterator<P, T> create(final int count, final int pageSize, final BiFunction<Integer, Integer, P> pageFetcher, Function<P, List<T>> itemExtractor, Function<P, Integer> totalPagesExtractor) {
+        return new PageSpliterator<>(0, count, pageSize, pageFetcher, itemExtractor, totalPagesExtractor);
+    }
+
     @Override
     public Spliterator<T> trySplit() {
-        if (pageSize * (pageNumber + 1) >= count) {
+
+        if (isLastPageOrGreater()) {
             return null;
         }
 
-        ChildPageSpliterator<T> childSpliterator = new ChildPageSpliterator<>(pageNumber, pageSize, pageFetcher);
-        this.pageNumber++;
+        if (preFetchedPage != null && getPageNumber() == 0) {
+            ChildPageSpliterator<List<T>, T> preFetchChildSpliterator = new ChildPageSpliterator<>(getPageNumber(), getPageSize(), (p, s) -> preFetchedPage, Function.identity());
+            incrementPageNumber();
+            return preFetchChildSpliterator;
+        }
+        ChildPageSpliterator<P, T> childSpliterator = new ChildPageSpliterator<>(getPageNumber(), getPageSize(), pageFetcher, itemExtractor);
+        incrementPageNumber();
         return childSpliterator;
     }
 
     @Override
     public long estimateSize() {
-        return count;
+        if (totalPagesExtractor != null && !hasPrefetched) {
+            prefetchPage();
+        }
+
+        if (totalNumberOfPages != null) {
+            return Math.min(getCount(), (long) getPageSize() * totalNumberOfPages);
+        }
+        return getCount();
```

Comment:
Generally not a fan of calling overridable methods unless its an important part of the abstraction.

Reply:
yea. makes sense. I had this as an abstract class at one point and never removed it